### ISSUE TITLE
Grain could stuck until timeout, when stream subscriber is not reachable (disconnected)

### ISSUE THREAD

**Issue by attilah** [dotnet/orleans#4197]
Problem:
I'm facing a streaming issue. I configured a SimpleMessageStream with Memory storage for pubsub. I'm having a simple client who instantiates a Grain, invoking a method on it, and inside the method the Grain calls ```OnNext``` on the stream to notify the subscribers of a state change.

If the client does not call ```UnsubscribeAsync``` before exiting, then on the next run, any call to the same Grain would timeout since the Grain is still "executing" the previous ```OnNext``` call.

It seems that client disconnection is not wiping out all the related data for that client, like subscription handles.

Solution:
a. When a client disconnects, all related data like subscription handles should be removed "immediately".
b. Check out why an ```OnNext``` to a disconnected client making the client stuck, why is there a wait for the timeout. Since client socket does not exists, so that remote callback to the client should fail much faster.

The error I saw in the console:
```
Runtime.Scheduler.WorkerPoolThread/5[101212]
      Worker pool thread Runtime.Scheduler.WorkerPoolThread/5 (ManagedThreadId=31) has been busy for long time: Executing Task Id=1 Status=Running for 00:05:00.1723371 on WorkItem=System*WorkItemGroup:Name=*stg/20/00000014,WorkGroupStatus=Running Executing for 00:05:00.1729365. WorkItemGroup Details: System*WorkItemGroup:Name=*stg/20/00000014,WorkGroupStatus=Running. Currently QueuedWorkItems=0; Total EnQueued=4; Total processed=3; Quantum expirations=0; TaskRunner=ActivationTaskScheduler-1:Queued=0; Detailed SchedulingContext=<[SystemTarget: S127.0.0.1:11111:258443880*stg/20/00000014@S00000014] CurrentlyExecuting=null>.
```

**Comment by xiazen** [comment]
> When a client disconnects, all related data like subscription handles should be removed "immediately".

This is not the expected behavior for client in streaming right now. In current streaming design, client is just another producer or consumer, it should maintain its own subscribe/unsubscribe or producing logic. Whether it should be or not is a question of what's the best behavior for streaming. And also doing what you say requires big structure change to streaming, because one of the fundamental assumption for stream consuming side is:  it is the consumer's responsibility to know when and which stream it want to subscribe/unsubscribe to. The streaming is just responsible for deliver messages on a stream to the consumers if they subscribed. 

In short, this is not a bug, but a design disagreement. If you think we should go for that design, mind elaborate more on your reasoning?

**Comment by xiazen** [comment]
>  Check out why an OnNext to a disconnected client making the client stuck, why is there a wait for the timeout. Since client socket does not exists, so that remote callback to the client should fail much faster.

I'm a bit confused, if the client is disconnected, how is it stuck? what do you mean by stuck?

If a producer is trying to produce message to a disconnected client, 
- sms stream would unregister the client from the stream at the first `ClientNotAvailableException` the producer received when trying to deliver message after the client disconnected
- persistent stream would also do the same. 

So what you described is not expected behavior. I modified the test `StreamConsumerOnDroppedClientTest` to see if I can reproduce what you are seeing. So I basically make the producer grain keep producing to the client after it is killed.  But I cannot. The behavior I see is, client got unregisted to the stream at the first `ClientNotAvailableException` the producer or the pulling agent received when trying to deliver itemt to it. 

So this maybe a bug , it is just I have trouble to reproduce it. So I'd like more information on this bug you see. What stream provider you are using? is it one of ours or a customized one? Mind share some logs to show what do you mean by stuck?

**Comment by jason-bragg** [comment]
> If the client does not call UnsubscribeAsync before exiting, then on the next run, any call to the same Grain would timeout since the Grain is still "executing" the previous OnNext call.

The expected behavior for clients that have become disconnected from the cluster (either by networking issues or ungraceful shutdown) is for the cluster to wait a configured period of time before considering the client dead, and cleaning up all date associated with that client after that period.  This period is the SiloMessagingOptions.ClientDropTimeout and defaults to 1 minute.  The weight time is to allow for transient networking errors.

This means that the onNext may fail for up to 1 minute after a client with a stream consumer disconnects from the cluster.  Is this the behavior you are seeing?

If so, this is expected behavior.  Some possible mitigations are:

* Reduce ClientDropTimeout.
* Add retry logic to grain.
* If message loss in error cases is acceptable, configure SMS to be fire and forget.  SimpleMessageStreamProviderOptions.FireAndForgetDelivery
